### ISSUE TITLE
CA1416 with guard attributes when non-taken path has UnsupportedOSPlatform attribute

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/roslyn-analyzers#7661]
### Analyzer

**Diagnostic ID**: warning CA1416: This call site is reachable on: 'macOS/OSX' 12.0 and later. 'TestType._DictionaryRepresentation13' is unsupported on: 'macOS/OSX' all versions.

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

```
$ ./.dotnet/dotnet --version
10.0.100-preview.3.25167.3
```

### Describe the bug

Incorrect CA1416 when trying to access API in code block protected using guard attributes, and the path not taken has code with UnsupportedOSPlatform attributes.

```cs
using System;
using System.Runtime.Versioning;

[assembly: SupportedOSPlatform ("macos12.0")]

partial class TestType {
    [SupportedOSPlatform ("macos12.0")]
    public virtual object DictionaryRepresentation {
        get {
            if (IsAtLeastXcode12) {
                return _DictionaryRepresentation14;
            } else {
                return _DictionaryRepresentation13; // warning CA1416: This call site is reachable on: 'macOS/OSX' 12.0 and later. 'TestType._DictionaryRepresentation13' is unsupported on: 'macOS/OSX' all versions.
            }
        }
    }

    [SupportedOSPlatformGuard ("macos")]
    public ulong? _DictionaryRepresentation14 { get; private set; }

    [Mock.UnsupportedOSPlatform ("macos")]
    public ulong? _DictionaryRepresentation13 { get; private set; }

    [SupportedOSPlatformGuard ("macos12.0")]
    internal static bool IsAtLeastXcode12 {
        get {
            return true;
        }
    }
}
```

Or as a unit test: https://github.com/rolfbjarne/roslyn-analyzers/commit/48ebc77ea21e147c8e0d6c69bbbb57f9f490274c

Reports:

### Steps To Reproduce

Run this unit test: https://github.com/rolfbjarne/roslyn-analyzers/commit/48ebc77ea21e147c8e0d6c69bbbb57f9f490274c

### Expected behavior

Unit test should pass.

### Actual behavior

Unit test fails:

```
System.InvalidOperationException : Context: Diagnostics of test state
Mismatch between number of diagnostics returned, expected "0" actual "1"

Diagnostics:
// /0/Test0.cs(16,24): warning CA1416: This call site is reachable on: 'ios' 14.0 and later, 'maccatalyst' 14.0 and later, 'macOS/OSX' 12.0 and later. 'TestType._DictionaryRepresentation13' is only supported on: 'ios' 12.2 and later, 'maccatalyst' 12.2 and later.
VerifyCS.Diagnostic(PlatformCompatibilityAnalyzer.CA1416).WithSpan(16, 24, 16, 51).WithArguments("TestType._DictionaryRepresentation13", "'ios' 12.2 and later, 'maccatalyst' 12.2 and later", "'ios' 14.0 and later, 'maccatalyst' 14.0 and later, 'macOS/OSX' 12.0 and later"),
```

## Additional context

This is not fixed by https://github.com/dotnet/roslyn-analyzers/pull/7569 nor https://github.com/dotnet/roslyn-analyzers/pull/7644.
